services/placement/selector/mirrored.go

```diff
@@ -0,0 +1,465 @@
+	for _, instance := range p.Instances() {
+		if _, ok := ssIDs[instance.ShardSetID()]; !ok {
+			continue
+		}
+		if instance.Hostname() == h.name {
+			continue
+		}
+
+		conflictRacks[instance.Rack()] = struct{}{}
+	}
+
+	var groups [][]services.PlacementInstance
+	for _, candidateHost := range hosts {
+		if candidateHost.name == h.name {
+			continue
+		}
+
+		if _, ok := conflictRacks[candidateHost.rack]; ok {
+			continue
+		}
+
+		groups, err = groupInstancesByHostPort([][]host{[]host{h, candidateHost}})
+		if err != nil {
+			f.logger.Warnf("could not match up candidate host %s with target host %s: %v", candidateHost.name, h.name, err)
+			continue
+		}
+
+		// Successfully grouped candidate with the host in placement.
+		break
+	}
+
+	if len(groups) == 0 {
+		return nil, errNoValidMirrorInstance
+	}
+
+	// The groups returned from the groupInstances() might not be the same order as
+	// the instances in leavingInstanceIDs. We need to reorder them to the same order
+	// as leavingInstanceIDs.
+	var res = make([]services.PlacementInstance, len(groups))
+	for _, group := range groups {
+		if len(group) != 2 {
+			return nil, fmt.Errorf("unexpected length of instance group for replacement: %d", len(group))
+		}
+
+		idx := findIndex(leavingInstanceIDs, group[0].ID())
+		if idx == -1 {
+			return nil, fmt.Errorf("could not find instance id: '%s' in leaving instances", group[0].ID())
+		}
+
+		res[idx] = group[1].SetShardSetID(group[0].ShardSetID())
+	}
+
+	return res, nil
+}
+
+func findIndex(ids []string, id string) int {
+	for i := range ids {
+		if ids[i] == id {
+			return i
+		}
+	}
+	// Unexpected.
+	return -1
+}
+
+func groupHostsByWeight(candidates []services.PlacementInstance) (map[uint32][]host, error) {
+	var (
+		uniqueHosts      = make(map[string]host, len(candidates))
+		weightToHostsMap = make(map[uint32][]host, len(candidates))
+	)
+	for _, instance := range candidates {
+		hostname := instance.Hostname()
+		weight := instance.Weight()
+		h, ok := uniqueHosts[hostname]
+		if !ok {
+			h = newHost(hostname, instance.Rack(), weight)
+			uniqueHosts[hostname] = h
+			weightToHostsMap[weight] = append(weightToHostsMap[weight], h)
+		}
+		err := h.addInstance(instance.Port(), instance)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return weightToHostsMap, nil
+}
+
+// groupHostsWithRackCheck looks at the racks of the given hosts
+// and try to make as many groups as possible. The hosts in each group
+// must come from different racks.
+func groupHostsWithRackCheck(hosts []host, rf int) ([][]host, []host) {
+	if len(hosts) < rf {
+		// When the number of hosts is less than rf, no groups can be made.
+		return nil, hosts
+	}
+
+	var (
+		uniqRacks = make(map[string]*rack, len(hosts))
+		rh        = racksByNumHost(make([]*rack, 0, len(hosts)))
+	)
+	for _, h := range hosts {
+		r, ok := uniqRacks[h.rack]
+		if !ok {
+			r = &rack{
+				rack:  h.rack,
+				hosts: make([]host, 0, rf),
+			}
+
+			uniqRacks[h.rack] = r
+			rh.Push(r)
```

Comment:
I think you should consistently use `heap.Push(rh, r)` instead of member methods, which is what the `container/heap` package example suggests. Can you do a sweep across the PR and make sure that's the case?